File: Sliding_Window/438_Find_All_Anagrams_in_String/solution438.cpp
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int startIndex = 0;
        unordered_map<char, int> pMap, sMap;
        vector<int> res;

        for (char c : p) {
            pMap[c]++;
        }

        for (int i = 0; i < s.length(); ++i) {
            sMap[s[i]]++;

            if (i >= p.length() - 1) {
                if (sMap == pMap) {
                    res.push_back(startIndex);
                }

                if (sMap[s[startIndex]] > 1) {
                    sMap[s[startIndex]]--;
                } else {
                    sMap.erase(s[startIndex]);
                }
                
                startIndex++;
            }
        }

        return res;
    }
};

// Runtime1: 38ms, beats 18.97%
// Memory1: 15.41MB, beats 18.19%

// Runtime2: 36ms, beats 21.63%
// Memory2: 15.33MB, beats 20.49%

// Runtime3: 33ms, beats 25.95%
// Memory3: 15.13MB, beats 21.84%

File: Sliding_Window/438_Find_All_Anagrams_in_String/solution438.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        
        startIndex = 0
        pMap, sMap = {}, {}
        res = []
        
        for char in p:
            pMap[char] = 1 + pMap.get(char, 0)
        
        for i in range(len(s)):
            sMap[s[i]] = 1 + sMap.get(s[i], 0)

            if i >= len(p) - 1:
                if sMap == pMap:
                    res.append(startIndex)
                
                if s[startIndex] in sMap:
                    sMap[s[startIndex]] -= 1
                    if sMap[s[startIndex]] == 0:
                        del sMap[s[startIndex]]
                startIndex += 1
        
        return res

# Runtime1: 89ms, beats 62.35%
# Memory1: 17.57MB, beats 96.49%

# Runtime2: 98ms, beats 52%
# Memory2: 17.62MB, beats 83.55%

# Runtime3: 77ms, beats 83.43%
# Memory3: 17.50MB, beats 99.60%